File: tp2/code/src/instance_reader.cpp
#include <iostream>
#include <sstream>
#include <fstream>
#include <cmath>

#include "instance_reader.hpp"

using namespace std;

InstanceReader::InstanceReader(string filename) {
    this->filename = filename;
}

int InstanceReader::calculate_euclidean_distance(point_t u, point_t v) {
    double dx = u.first - v.first;
    double dy = u.second - v.second;

    return round(sqrt(dx*dx + dy*dy));
}

int InstanceReader::calculate_pseudo_euclidean_distance(point_t u, point_t v) {
    double dx = u.first - v.first;
    double dy = u.second - v.second;

    double real_dist = sqrt((dx*dx + dy*dy) / 10.0);
    int rounded_dist = round(real_dist);

    if (rounded_dist < real_dist)
        return rounded_dist + 1;

    return rounded_dist;
}

graph_t InstanceReader::read_instance() {
    string line;
    ifstream file(this->filename);
    
    // Lendo cada linha do arquivo até NODE_COORD_SECTION salvando as dimensões e tipo de peso
    while (getline(file, line)) {
        if (line == "NODE_COORD_SECTION")
            break;

        string keyword, value;
        stringstream stream(line);

        stream >> keyword >> value;
        if (keyword == "NAME:")
            this->instance_name = value;

        else if (keyword == "TYPE:") {
            if (value != "TSP") {
                cout << "error: instance type must be TSP" << endl;
                exit(1);
            }
        }

        else if (keyword == "DIMENSION:")
            this->dimension = stoi(value);

        else if (keyword == "EDGE_WEIGHT_TYPE:") {
            if (value != "EUC_2D" && value != "ATT") {
                cout << "error: edge weight type must be either EUC_2D or ATT" << endl;
                exit(1);
            }

            this->edge_weight_type = value;
        }
    }

    // Lendo cada coordenada da nossa instância
    coords_t coords(this->dimension);
    for (int i = 0; i < this->dimension; i++) {
        int id;
        double x, y;

        getline(file, line);
        stringstream stream(line);

        stream >> id >> x >> y;
        coords[i] = make_pair(x, y);
    }

    // Criando grafo baseado nas coordenadas e tipo de peso das arestas
    graph_t graph(this->dimension, vector<double>(this->dimension));
    for (int i = 0; i < this->dimension; i++) {
        for (int j = i+1; j < this->dimension; j++) {
            if (this->edge_weight_type == "EUC_2D")
                graph[i][j] = graph[j][i] = this->calculate_euclidean_distance(coords[i], coords[j]);

            else
                graph[i][j] = graph[j][i] = this->calculate_pseudo_euclidean_distance(coords[i], coords[j]);   
        }
    }

    // Definindo a diagonal da matriz como infinito (não há aresta de um vértice para ele mesmo)
    for (int i = 0; i < this->dimension; i++)
        graph[i][i] = INF;

    file.close();
    return graph;
}

File: tp2/utils/instance_reader.py
import numpy as np
from typing import List

from .types import Point
from .distances import calculate_euclidean_distance
from .distances import calculate_pseudo_euclidean_distance


class InstanceReader:
    def __init__(self, filename: str) -> None:
        with open(filename, 'r') as f:
            line = f.readline()

            while line.strip() != "NODE_COORD_SECTION":
                keyword, value = line.split(':')
                value = value.strip()

                if keyword == "NAME":
                    self.instance_name = value
                
                elif keyword == "DIMENSION":
                    self.dimension = int(value)
                
                elif keyword == "EDGE_WEIGHT_TYPE":
                    self.distance_type = value
                    self.distance = calculate_pseudo_euclidean_distance if value == "ATT" else calculate_euclidean_distance

                line = f.readline()

            # Anotando as coordenadas
            self.coords: List[Point] = []
            for _ in range(self.dimension):
                _, x, y = f.readline().split()
                self.coords.append(Point(float(x), float(y)))

    def build_complete_graph(self) -> np.ndarray:
        graph = np.zeros((self.dimension, self.dimension))

        for i in range(self.dimension):
            for j in range(i+1, self.dimension):
                graph[i][j] = self.distance(self.coords[i], self.coords[j])
                graph[j][i] = graph[i][j]

        np.fill_diagonal(graph, np.inf)
        return graph
